kratos/future/solving_strategies/builders/block_builder.h

```diff
@@ -0,0 +1,375 @@
+
+            // Apply constraints to LHS
+            auto p_lhs = rLinearSystemContainer.pLhs;
+            auto p_LHS_T = AmgclCSRSpMMUtilities::SparseMultiply(*p_lhs, *rLinearSystemContainer.pEffectiveT);
+            auto p_transT = AmgclCSRConversionUtilities::Transpose(*rLinearSystemContainer.pEffectiveT);
+            rLinearSystemContainer.pEffectiveLhs = AmgclCSRSpMMUtilities::SparseMultiply(*p_transT, *p_LHS_T);
+
+            // // Compute the scale factor value
+            // //TODO: think on how to make this user-definable
+            // const double scale_factor = rpEffectiveLhs->NormDiagonal();
+
+            // // Apply diagonal values on slave DOFs
+            // IndexPartition<IndexType>(mSlaveIds.size()).for_each([&](IndexType Index){
+            //     const IndexType slave_eq_id = mSlaveIds[Index];
+            //     if (mInactiveSlaveDofs.find(slave_eq_id) == mInactiveSlaveDofs.end()) {
+            //         (*rpEffectiveLhs)(slave_eq_id, slave_eq_id) = scale_factor;
+            //         (*rpEffectiveRhs)[slave_eq_id] = 0.0;
+            //     }
+            // });
+        } else {
+            // If there are no constraints the effective arrays are the same as the input ones
+            // Note that we avoid duplicating the memory by making the effective pointers to point to the same object
+            rLinearSystemContainer.pEffectiveLhs = rLinearSystemContainer.pLhs;
+            rLinearSystemContainer.pEffectiveRhs = rLinearSystemContainer.pRhs;
+            rLinearSystemContainer.pEffectiveDx = rLinearSystemContainer.pDx;
+        }
+    }
+
+    /**
+     * @brief Applies the Dirichlet conditions to the system
+     * This method applies the Dirichlet conditions following a block-type build
+     * @param rDofArray DOFs array from which the fixity is checked
+     * @param rLHS Left Hand Side matrix
+     * @param rRHS Right Hand Side vector
+     */
+    void ApplyBlockBuildDirichletConditions(
+        const DofsArrayType& rDofArray,
+        TSparseMatrixType& rLHS,
+        TSystemVectorType& rRHS) const
+    {
+        // Set the free DOFs vector (0 means fixed / 1 means free)
+        // Note that we initialize to 1 so we start assuming all free
+        // Also note that the type is uint_8 for the sake of efficiency
+        const std::size_t system_size = rLHS.size1();
+        std::vector<uint8_t> free_dofs_vector(system_size, 1);
+
+        // Loop the DOFs to find which ones are fixed
+        // Note that DOFs are assumed to be numbered consecutively in the block building
+        const auto dof_begin = rDofArray.begin();
+        IndexPartition<std::size_t>(rDofArray.size()).for_each([&](IndexType Index){
+            const auto p_dof = dof_begin + Index;
+            if (p_dof->IsFixed()) {
+                free_dofs_vector[p_dof->EffectiveEquationId()] = 0;
+            }
+        });
+
+        //TODO: Implement this in the CSR matrix or here?
```

Comment:
This is not done the next line with diagonal value?

Reply:
Indeed, thanks for noting this as this is something that we wanted to discuss too.